src/libs/saleforce.ts

```diff
@@ -0,0 +1,171 @@
+}
+
+export interface SalesforceBearerInformation {
+    access_token: string;
+    instance_url: string;
+}
+
+interface PhoneBookQueryAnswerItem {
+    Name: string;
+    Buyer__r: {
+        Phone: string | null;
+        IdentityID__c: string | null;
+    };
+}
+
+interface PhoneBookQueryAnswerData {
+    records: PhoneBookQueryAnswerItem[];
+}
+
+export interface PhoneRecord {
+    subscriptionName: string;
+    phoneNumber: string | null;
+    identityId: string | null;
+}
+
+export type PhoneBook = PhoneRecord[];
+
+export async function makeSalesforceSSMConfig(
+    stage: string,
+): Promise<SalesforceSSMConfig> {
+    const clientId = await getSsmValue(stage, 'salesforceClientId');
+    const clientSecret = await getSsmValue(stage, 'salesforceClientSecret');
+    const username = await getSsmValue(stage, 'salesforceUsername');
+    const password = await getSsmValue(stage, 'salesforcePassword');
+    const token = await getSsmValue(stage, 'salesforceToken');
+    const authenticationBaseUrl = await getSsmValue(
+        stage,
+        'salesforceAuthenticationBaseUrl',
+    );
+    if (
+        clientId === undefined ||
+        clientSecret === undefined ||
+        username === undefined ||
+        password === undefined ||
+        token === undefined ||
+        authenticationBaseUrl == undefined
+    ) {
+        throw 'Could not retrive the salesforce ssm config';
+    }
+    return Promise.resolve({
+        clientId,
+        clientSecret,
+        username,
+        password,
+        token,
+        authenticationBaseUrl,
+    });
+}
+
+async function getSalesforceBearerInformation(
+    saleforceSSMConfig: SalesforceSSMConfig,
+): Promise<SalesforceBearerInformation> {
+    console.log('Querying bearer token from Salesforce');
+
+    const requestBody =
+        `grant_type=password` +
+        `&client_id=${saleforceSSMConfig.clientId}` +
+        `&client_secret=${saleforceSSMConfig.clientSecret}` +
+        `&username=${saleforceSSMConfig.username}` +
+        `&password=${saleforceSSMConfig.password}${saleforceSSMConfig.token}`;
+
+    const url = `${saleforceSSMConfig.authenticationBaseUrl}/services/oauth2/token`;
+
+    try {
+        const params = {
+            headers: {
+                'Content-Type': 'application/x-www-form-urlencoded',
+            },
+        };
+        const response = await axios.post(url, requestBody, params);
+        return (await response.data) as SalesforceBearerInformation;
+    } catch (error) {
+        throw new Error(
+            `error while retrieving salesforce bearer token: ${error}`,
+        );
+    }
+}
+
+async function runPhoneBookQuery(
+    bearerInformation: SalesforceBearerInformation,
+): Promise<PhoneBookQueryAnswerData> {
+    console.log('Running phone book query');
+
+    const query =
+        "SELECT Name, Buyer__r.IdentityID__c, Buyer__r.Phone FROM SF_Subscription__c WHERE Product__c = 'Newspaper - National Delivery'";
```

Comment:
This may need to be filtered more carefully at some point in the future. Once it approaches 2000 records returned, the results may be paginated (see `nextRecordsUrls` [here](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/dome_query.htm)). 

Currently less than 500 results are returned, but running the equivalent query for the "Home Delivery" (inside M25) product as an example more than 18,000 subs are returned.

This might be a problem for another day, but some ideas to add to the WHERE clause might be:

- add `SF_Status__c IN ('Active', 'Cancellation Pending')` to restrict to only currently active subscriptions
- add `Buyer__r.Phone != null` to restrict to only subscriptions with buyers with phone numbers

Adding both to a 'Home Delivery' query reduces the results from over 18,000 down to just under 5,000. So eventually we'll still hit the pagination limit so that just might be something we have to deal with eventually.

Reply:
Thanks for pointing all of this out Graham. Definitively a health card opportunity.

Reply:
https://trello.com/c/D18uWM9k/128-improve-the-way-phone-numbers-are-queried-in-the-fulfilment-files